### PR TITLE
[IOPID-1254] : Audit Log

### PR DESCRIPTION
**Pull request by ClaudioNTT**
### Summary

The proposed modification introduces the audit log within the BFF using the "@azure/storage-blob" library as requested
The choice of this library has been made to enable the usage of tags

**NOTE:**
For this modification, the APIM needs to be adjusted to allow the client's IP to be redirected in the header

#### List of Changes

- Created the utility for writing audit logs
- Introduced adutilog into the exchange chain

#### Motivation and Context
It has been introduced as requested by task [IOPID-1254](https://pagopa.atlassian.net/browse/IOPID-1254)

#### How Has This Been Tested?

- Installed Azurite in the development environment to test the functionality as per Microsoft documentation
- Adjusted unit tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

[IOPID-1254]: https://pagopa.atlassian.net/browse/IOPID-1254?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Exchange/handler.ts [hunk 1]
```diff
@@ -61,5 +120,5 @@ export const getExchangeHandler = (config: IConfig): express.RequestHandler => {
     )
   );
 
-  return wrapRequestHandler(middlewaresWrap((_, user) => handler(user)));
+  return wrapRequestHandler(middlewaresWrap((_, user) => handler(user, _)));
```
Comment (gquadrati) [inline Exchange/handler.ts]:
```suggestion
  return wrapRequestHandler(
    middlewaresWrap((context, user) => handler(user, context))
  );
```

File: utils/audit-log.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+): TE.TaskEither<RestError, BlockBlobUploadResponse> =>
+  TE.tryCatch(
+    () => {
+      const blobServiceClient = BlobServiceClient.fromConnectionString(
+        config.AUDIT_LOG_CONNECTION_STRING
+      );
+      const containerClient = blobServiceClient.getContainerClient(
+        config.AUDIT_LOG_CONTAINER
+      );
+
```
Comment (gquadrati) [inline utils/audit-log.ts]: We don't need to create a blobServiceClient each time, we can define it in `index.ts` and pass it as dependency to the handler up to this function. See an example [here](https://github.com/pagopa/io-functions-fast-login/blob/8fad02e9951b2ca94b6e7f2925c9dcc3e0522b4e/src/main.ts#L44)

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: Exchange/handler.ts [hunk 3]
```diff
@@ -45,6 +72,38 @@ export const exchangeHandler = (config: IConfig): ExchangeHandlerT => (
...
+                    context.req.headers["x-forwarded-client-ip"]
+                  ) as IPString)
+                : ("" as IPString),
+              tokenId: decodedToken.jti,
+              tokenIssuingTime: new Date(decodedToken.iat * 1000).toISOString()
+            },
+            {
+              DateTime: new Date(decodedToken.iat * 1000).toISOString(),
+              FatherIDToken: user_data.jti,
+              FiscalCode: decodedToken.fiscal_number,
```
Comment (gquadrati) [inline Exchange/handler.ts]: We need to hash it. We can use `hashFiscalCode` function from `@pagopa/ts-commons` last version.

File: utils/audit-log.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+export const generateBlobName = (
+  fiscal_number: FiscalCode,
+  token_type: TokenTypes,
+  token_id: string
+): string => {
+  const UTCDateTime = new Date().toISOString();
+  const hash = createHash("md5")
+    .update(fiscal_number)
+    .digest("hex");
```
Comment (gquadrati) [inline utils/audit-log.ts]: Ditto: we can use `hashFiscalCode` function from `@pagopa/ts-commons` last version.

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: Exchange/handler.ts (hunk 3 above)
Comment (ClaudioNTT) [reply Exchange/handler.ts]: done with last commit, thanks!

File: utils/audit-log.ts (hunk 4 above)
Comment (ClaudioNTT) [reply utils/audit-log.ts]: done with last commit, thanks!

File: Exchange/handler.ts [hunk 5]
```diff
@@ -45,12 +76,47 @@ export const exchangeHandler = (config: IConfig): ExchangeHandlerT => (
...
+        decodeToken(jwt),
+        TE.chain(decodedToken =>
+          storeAuditLog(
+            containerClient,
+            {
+              ip: context.req
+                ? (JSON.stringify(
+                    context.req.headers["x-forwarded-client-ip"]
+                  ) as IPString)
+                : ("" as IPString),
```
Comment (gquadrati) [inline Exchange/handler.ts]: We can relax the ip type to string

```suggestion
             ip: context.req?.headers["x-forwarded-client-ip"] ?? "",
```

**Review by gquadrati — APPROVED** [review]

**Pull request merged**